### PR TITLE
Remove support for HASH partitioning

### PR DESCRIPTION
**Pull request by danielgustafsson**
Hash partitioning was never fully implemented, and was never turned on by default. There has been no effort to complete the feature, so rather than carrying dead (or perhaps limping) code this removes all support for hash partitioning. Should we ever want this feature, we will most likely start from scratch anyways.

As an effect from removing the unsupported `MERGE`/`MODIFY` commands, this previously accepted query is no longer legal:

```SQL
create table t (a int, b int)
distributed by (a)
partition by range (b) (start () end(2));
```

The syntax was an effect of an incorrect rule in the parser which made the start boundary optional for `CREATE TABLE` when it was only intended for `MODIFY PARTITION`. We only had a single test for this which was commented "legal?!" so I doubt it's seen much use. Iff there is usage of this we might have to reimplement this support, but then it should at least be defined/documented what the query will actually do.

pg_upgrade was already checking for hash partitions so no new check was required (upgrade would've been impossible anyways due to hash algorithm change).

Two smaller cleanup commits are also included in here as they fell out while poking at this.

### CODE REVIEW

File: contrib/pg_upgrade/check.c [hunk 1]
```diff
@@ -993,13 +993,13 @@ check_covering_aoindex(migratorContext *ctx)
...
  *	check_hash_partition_usage()
  *	8.3 -> 8.4
- *	The hash algorithm was changed in 8.4, so upgrading is impossible. This
- *	is basically the same problem as with hash indexes in PostgreSQL. Hash
- *	partitioning was not officially supported in GPDB5, but better check just
- *	in case someone has found the hidden GUC and used them anyway.
  *
- *	XXX: Actually, pg_dump outright fails on hash partitioned tables, so we
- *	cannot support hash partitioned tables even on a same-version upgrade.
+ *	Hash wever never officially supported in GPDB5 and was removed in GPDB6,
```
Comment (vraghavan78) [inline contrib/pg_upgrade/check.c]: wever -> was

File: src/test/regress/expected/partition.out [hunk 2]
```diff
@@ -1,39 +1,79 @@
 set enable_partition_rules = false;
 set gp_enable_hash_partitioned_tables = true;
```
Comment (vraghavan78) [inline src/test/regress/expected/partition.out]: can we remove this GUC>

File: src/test/regress/sql/partition.sql [hunk 3]
```diff
@@ -2,18 +2,30 @@ set enable_partition_rules = false;
 set gp_enable_hash_partitioned_tables = true;
```
Comment (vraghavan78) [inline src/test/regress/sql/partition.sql]: remove above line

**Comment by danielgustafsson** [comment]
> On 13 Nov 2017, at 20:00, Venkatesh Raghavan <notifications@github.com> wrote:
> 
> @vraghavan78 commented on this pull request.
> 
> In src/test/regress/expected/partition.out:
> 
> > @@ -1,39 +1,79 @@
>  set enable_partition_rules = false;
>  set gp_enable_hash_partitioned_tables = true;
> 
> can we remove this GUC>

It has been removed, I left the (failing) test for it to ensure it wasn’t accidentally re-enabled by a mismerge at some point in the near future.

**Comment by danielgustafsson** [comment]
Pushed to master following a green CI run

**Pull request closed without merging**